11-golden_quail/technical_specification.md

```diff
@@ -0,0 +1,71 @@
+# Expanding Hexkit to Abstract the Persistence Layer (Golden Quail)
+**Epic Type:** Implementation Epic
+
+**Attention: Please do not put any confidential content here.**
+
+## Scope:
+A scope definition can be found here: https://wiki.verbis.dkfz.de/x/uQAECg
+
+The following documents might be helpful to understand the context and objective of this epic:
+1. [Hexagonal Architecture Design Pattern](https://wiki.verbis.dkfz.de/x/noAlCg)
+2. [Triple Hexagonal Architecture Concept](https://wiki.verbis.dkfz.de/x/MABFCg)
+3. [Reducing Redundancy and Ensuring Consistency between Microservices](https://wiki.verbis.dkfz.de/x/tgFVCg)
+
+
+## Implementation Details:
+
+This epic shall add outbound protocol/provider pairs for interacting with persistence architecture including databases
+and object storages. The functionality implemented in the hexkit at the end of this step is summarized in the following figure:  
+
+![](./images/protocol_and_providers_overview.jpg)
+
+### S3-based Object Storage:
+An abstraction around S3-like object storages is already provided in the ghga-service-chassis-lib.
+Thereby the [ObjectStorageDao](https://github.com/ghga-de/ghga-service-chassis-lib/blob/main/ghga_service_chassis_lib/object_storage_dao.py#L259) roughly corresponds to a protocol and the [ObjectStorageS3](https://github.com/ghga-de/ghga-service-chassis-lib/blob/main/ghga_service_chassis_lib/s3.py#L175) roughly corresponds to a provider implementation.
+
+These classes need to be ported to the hexkit and turned into async code.
+Since the underlying boto3 library is synchronous, blocking calls from that library will be wrapped with the
+`asyncio.to_thread` function to make them awailable.
+
+### Basic CRUD Database:
+A protocol for basic CRUD interactions with a database will be proposed. The operations shall include:
+- get resource by ID
+- create resource
+- update resource
+- upsert resource (if resource exists, will update, otherwise will create)
+- delete resource by ID
+- find resources by matching one of their attributes to a specific value
```

Comment:
A list of key-value pairs is definitely possible and a good idea. However, I am not sure about conditions. With that we will easily create our own querying language, which is something I would like to avoid 😆

Reply:
Sorry, I did not mean to add arbitrary conditions, just that the list of key-value pairs should act as filter where *all* conditions (key=value) must be met ("and" condition).

Reply:
done